dsl/pace/dsl/dace/doc_primer_orchestration.md

```diff
@@ -0,0 +1,101 @@
+DaCe Orchestration in Pace: a primer
+====================================
+
+
+Fundamentals
+------------
+
+Full program optimziation with DaCe is the process of turning all Python and GT4Py code in generated C++.
+
+_Orchestration_ is the our own wording for full program optimization. We only _orchestrate_ the runtime code of the model, e.g. everything in the `__call__` method of the module. All code in `__init__` is executed like a normal gt backend.
+
+At the highest level in Pace, to turn on orchestration you need to flip the `FV3_DACEMODE` to an orchestrated options _and_ run a `dace:*` backend (it will error out if run anything else). Option for `FV3_DACEMODE` are:
+- _Python_: default, turns orchestration off.
+- _Build_: build the SDFG then exit without running. See Build for limitation of build strategy.
+- _BuildAndRun_: as above, but distribute the build and run.
+- _Run_: tries to execute, errors out if the cache don't exists.
+
+Code is orchestrated two ways:
+- functions are orchestrated via `orchestrate_function` decorator
+- methods are orchestrate via the `orchestrate` function (e.g. `pace.driver.Driver._critical_path_step_all`)
+
+The later is the way we orchestrate in our model. `orchestrate` is often called as the first function in the `__init__`. It patches _in place_ the methods and replace them with a wrapper that will deal with turning it all into executable SDFG when call time comes.
+
+The orchestration has two parameters: config (will expand later) and `dace_compiletime_args`.
+
+DaCe needs to be described all memory so it can interface it in the C code that will be executed. Some memory is automatically parsed (e.g. numpy, cupy, scalars) and others need description. In our case `Quantity` and others need to be flag as `dace.compiletime` which tells DaCe to not try to AOT the memory and wait for JIT time. The `dace_compiletime_args` helps with tagging those without having to change the type hint.
+
+File structure
+--------------
+
+`pace.dsl.dace.*` carries the structure for orchestration.
+* `build.py`: tooling for distributed build & SDFG load.
+* `dace_config.py`: DaCeConfig & DaCeOrchestration enum.
+* `orchestration.py`: main code, takes care of orchestration .scaffolding, build pipeline (including parsing) and execution.
+* `sdfg_opt_passes.py`: custom optimization pass for Pace, used in the build pipeline.
+* `utils.py`: as every "utils" or "misc" or "common" file, this should not exists and collect tools & functions I lazily didn't put in a proper place.
+* `wrapped_halo_exchange.py`: a callback-ready halo exchanger, which is our current solution for keeping the Halo Exchange in python (because of prior optimization) in orchestration
```

Comment:
```suggestion
* `wrapped_halo_exchange.py`: a callback-ready halo exchanger, which is our current solution for keeping the Halo Exchange in python (because of prior optimization) in orchestration.
```